tests/verify.rs

```diff
@@ -15,6 +22,35 @@ fn should_link_example_firmware() -> anyhow::Result<()> {
     Ok(())
 }
 
+#[test]
+fn should_verify_memory_layout() -> anyhow::Result<()> {
+    // Arrange
+    cargo::check_flip_link();
+
+    // Act
+    cargo::build_example_firmware(CRATE).success();
+
+    // Assert
+    for elf_path in elf::paths() {
+        // read and parse elf-file
+        let elf = fs::read(elf_path)?;
+        let object = object::File::parse(&*elf)?;
+
+        // get the relevant sections
+        let (bss, data, uninit, vector_table) = elf::get_sections(&object)?;
+        // compute the initial stack-pointer from `.vector_table`
+        let initial_sp = elf::compute_initial_sp(&vector_table)?;
+        // get the bounds of 'static RAM'
+        let bounds = elf::get_bounds(&[data, bss, uninit])?;
+
+        // Is the initial stack-pointer inside 'static RAM'?
+        assert!(bounds.contains(&initial_sp));
```

Comment:
This is not the correct thing to assert even though the test passes.

the `bounds` range span all static variables in RAM. `flip-link` will put the stack *right below* that. we want to check (assert) that the stack and the static variables do *not* overlap

the tricky bit is that `initial_sp` points to the start of the stack but when you use the stack *that* address will not be written to: data is written to the address immediately *below* that (this is because on ARM the stack is of the *full descending* type).

So if the SP is `0x2000_3000` and you push `0xAAAA_AAAA` (32-bit word) onto it the memory that is written is:
- `0x2000_3000` this address is not written to
- `0x2000_2FFF` = 0xAA
- `0x2000_2FFE` = 0xAA
- `0x2000_2FFD` = 0xAA
- `0x2000_2FFC` = 0xAA
- new SP value = `0x2000_2FFC`

On the other hand if you have a static variable located at `0x2000_3000` and write to it then address `0x2000_3000` will be written to. Thus `initial_sp` can be equal to `bounds.start()`.

so the cases here are
- `initial_sp <= bounds.start` = OK (stack is below static variables)
- `initial_sp > bounds.start && initial_sp < bounds.end` = BAD (overlap)
- `initial_sp > bounds.end` = BAD (stack is *above* static variables -> if the stack grows it could corrupt static variables)

Reply:
Ah kay. Thank you for the elaborate explanation! 😄

Reply:
I am thinking of putting this explanation either into a doc-comment or some kind of `CONTRIBUTING.md`, since I found it very useful. What do you think would be a good place for it?

Reply:
I tend to favor 'put it where it's hard to miss' and to me that's usually near the code that implements or verifies the thing because (hurried) devs tend to miss markdown files that are not the README (:raising_hand: ).

A separate markdown file would also be OK but then we need to link back from the related code ("before working on this module / function read `some-file.md`). Keeping that link up to date could become a chore -- though this code base is small and doesn't change much.